rfcs/20201103-binary-bcast-source0/README.md

```diff
@@ -0,0 +1,74 @@
+# Proposal for enable broadcast source 0 in binary
+
+## Problem and motivation
+Currently in OneDNN library binary primitive supports implicit broadcast 
+semantics for source 1. It means that if some dimension has value of one, 
+this value will be used to compute an operation with each point of 
+source 0 for this dimension. There is also a need for enabling this functionality
+for source 1 to handle binary operation on sources like `Nx1` and `1xM`. This 
+will improve oneDNN integration in TensorFlow library. At this point, was 
+reported, that performance (jit kernel) is not the main concern.
+
+![](bcast_example.png "Binary ADD operation 1x4:4x1")
+
+## Overview
+
+### Current implementations
+*  Reference implementation allows broadcast with following syntax:
+   `NxCxDxHxW:N/1xC/1xD/1xH/1xW/1 -> NxCxDxHxW`
+*  JIT implementation allows broadcast with following syntax:
+   `NxCxDxHxW:{NxCx1x1x1,1xCx1x1x1,1x1x1x1x1} -> NxCxDxHxW`
+*  JIT i8 implementation allows broadcast with following syntax:
+   `NxCxDxHxW:{N,1}xCx1x1x1 -> NxCxDxHxW`
+### New implementations
+*  Reference implementation allows broadcast with following syntax:
+   `N/1xC/1xD/1xH/1xW/1:N/1xC/1xD/1xH/1xW/1 -> NxCxDxHxW`
+### Possible future implementations (if the need is reported)
+*  JIT implementation allows broadcast with following syntax:
+   `{NxCxDxHxW,NxCx1x1x1,1xCx1x1x1,1x1x1x1x1}:
+    {NxCxDxHxW,NxCx1x1x1,1xCx1x1x1,1x1x1x1x1} -> NxCxDxHxW`
+*  JIT i8 implementation allows broadcast with following syntax:
+   `{N,1}xCx{D,1}x{H,1}x{W,1}:{N,1}xCx{D,1}x{H,1}x{W,1} -> NxCxDxHxW`
+
+Similarly for other formats.
+
+## Proposal
+The proposal is to extend binary primitive by removing a restriction 
+on src0 broadcasting. The reference implementation will be the only 
+implementation supporting this broadcasting until we have an additional 
+request. The dimensions of the destination should be defined as: 
+`dst_dims[d] = max(src0_dims[d], src1_dims[d])` and 
```

Comment:
Right now dst_md is allowed to have format_kind::any which further is converted into src0 memory format. Since dimensions will not match, what's the plan to handle format_kind::any for dst in such case?

Reply:
I added section for this problem with two solutions.